src/compiler/checker.ts

```diff
@@ -27093,6 +27096,12 @@ namespace ts {
             const elementTypes: Type[] = [];
             const elementFlags: ElementFlags[] = [];
             const contextualType = getApparentTypeOfContextualType(node);
+            if (elementCount === 0 && contextualType && !checkMode) {

```

Comment:
I think this may need to be moved lower, right below the tuple check - otherwise you'll never be able to get an empty tuple inferred via a contextual type.